### ISSUE TITLE
monitoring -  the app monitoring-operator fails to upgrade manually

### ISSUE THREAD

**Issue by jiaqiluo** [rancher/rancher#20301]
**What kind of request is this (question/bug/enhancement/feature request):**
Bug

**Steps to reproduce (least amount of steps as possible):**
- run Rancher: v2.2.3 container
- add an RKE EC2 cluster
- enable cluster monitoring 
- upgrade Rancher to v2.2.4-rc5
- go to project `system` - `Apps`
- upgrade the app `cluster-monitoring` up to date (v0.0.3)
- wait for `cluster-monitoring` to be active
- upgrade the app `monitoring-operator` up to date (v0.0.3)

**Result:**
- upgrading fails with the following error:

Failed to install app monitoring-operator. Error: UPGRADE FAILED: Deployment.apps "prometheus-operator-monitoring-operator" is invalid: spec.selector: Invalid value: v1.LabelSelector{MatchLabels:map[string]string{"app":"prometheus-operator", "chart":"prometheus-operator-0.0.1", "release":"monitoring-operator"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable

<img width="1376" alt="screenshot" src="https://user-images.githubusercontent.com/6218999/57895372-2229ab00-7800-11e9-8b27-36ae6cda3646.png">

**Comment by orangedeng** [comment]
We need to use this option when upgrading operator:
<img width="1169" alt="屏幕快照 2019-05-17 下午12 23 16" src="https://user-images.githubusercontent.com/15376837/57903002-a468ac80-789e-11e9-8fd7-42f30c324c86.png">

**Comment by orangedeng** [comment]
The root cause of the install error is that we move the operator charts from root template to sub charts template in version 0.0.3. And the version for the operator app changed from 0.0.2 to 0.0.1(which is bad because it is going backward). I would prefer to keep the 0.0.2 version to operator and the upgrade would be fine.

PS. The users won't be required to upgrade operator to version 0.0.3. And it won't affect monitoring if the user don't upgrade it.

**Comment by sangeethah** [comment]
@loganhz @orangedeng If you have an "Upgrade Available" options , users are going to likely attempt to upgrade the app.

Why do we not want to have the operator app version set to 0.0.3 , so that upgrade of "operator app" succeeds in this case ? 

The option to include "nameOvrride" option when upgrading is something that needs to be documented and is not an ideal option .

**Comment by loganhz** [comment]
After the PR merged, you will not see "Upgrade Available" options for operator

**Comment by cjellick** [comment]
Why not @loganhz?

**Comment by loganhz** [comment]
There is no change for operator chart.

**Comment by jiaqiluo** [comment]
@orangedeng   
On the same setup shown in the issue, I go to click `upgrade` for the `monitoring-operator` app, and the option is already there:
<img width="1431" alt="screenshot" src="https://user-images.githubusercontent.com/6218999/57947424-c3157600-7893-11e9-8e93-fd9359872af8.png">

I change nothing and click the `upgrade` button, then the app got deployed and be active. 

The same behavior happens on both the following upgrading path:
- v2.2.3 -> v2.2.4-rc5
- v2.2.1 -> v2.2.4-rc5

**Comment by loganhz** [comment]
Can you always reproduce it? I cannot reproduce it in my env. @jiaqiluo

**Comment by sangeethah** [comment]
@loganhz . Yes. We  saw this issue happen every time when tried upgrade scenario.

**Comment by alena1108** [comment]
@loganhz @sangeethah as we are planning to implement one click upgrade for monitoring (https://github.com/rancher/rancher/issues/20304), this issue will no longer be relevant for the upgrade use case.

**Comment by thxCode** [comment]
Since `Prometheus-Operator`, `Prometheus`, `Grafana` and `Alertmanager` group into one chart, when every new version has been published,  monitoring related rancher Apps (eg: `cluster-monitoring`, `monitoring-operator`, `cluster-alerting` and `project-monitoring` per Project) could see the `Upgrade Available` tip from UI.

- For `system-charts/rancher-monitoring:v0.0.2`, the `Prometheus-Operator` component is the main template in the chart, so the `0.0.2` is not only the chart's version but also the `Prometheus-Operator` component's **logical version** of the chart:
    ```
    rancher-monitoring/v0.0.2/
      charts/       # putting other components in here
      templates/    # putting Prometheus-Operator templates in here
      Chart.yaml    # 0.0.2 marks in this file
      ...
    ```
- For `system-charts/rancher-monitoring:v0.0.3`, the `Prometheus-Operator` component moves into the charts directory,  we can treat all components as sub-charts and use `xxx.enabled` to enable/disable `xxx` component, so the `0.0.3` is only the chart's version:
    ```
    rancher-monitoring/v0.0.3/
      charts/         # putting other components in here
        ...
        operator/
          templates/  # putting Prometheus-Operator templates in here
          Chart.yaml  # Operator's logic version marks in this file, for now, it's 0.0.1
      Chart.yaml      # 0.0.3 marks in this file
      ...
    ```
- Logically,

When upgrading to `monitoring-operator` from `v0.0.2` to `v0.0.3`, the `chart` label value (which generate via charts/sub-charts **logical version**)  in `spec.selector.matchLabels` has been changed from `prometheus-operator-0.0.1` to `prometheus-operator-0.0.2`. However `spect.selector` is an immutable field, we cannot modify this since `app/v1beta2`:  https://github.com/kubernetes/kubernetes/issues/50808. `chart` label value generates via chart version.

In general, in order to ensure forward compatibility (don't cause this issue again), we should ensure that the **logical version** in the sub-charts do not change, in fact, they should not be changed.

I think https://github.com/rancher/system-charts/pull/47 could fix this.

**Comment by loganhz** [comment]
@thxcode can you explain why users will still be able to see the upgrade available button if you don’t change the version here?

**Comment by loganhz** [comment]
So the change in the PR is for sub chart version. The sub chart version is only used for the label. So we don’t have to change it. Users still can see the upgrade button. But users don’t have to select force upgrade any more when they upgrade monitoring operator app

**Comment by alena1108** [comment]
Can be tested with `v2.2.4-rc6`

**Comment by jiaqiluo** [comment]
The bug fix validated on Rancher: 2.2.4-rc6

Test 1:
- run Rancher: v2.2.1 container
- add an RKE EC2 cluster
- enable cluster monitoring 
- add a new project `p1`
- enable project monitoring
- upgrade Rancher to v2.2.4-rc6
- go to project `system` - `Apps`
- upgrade the app `cluster-monitoring` up to date (v0.0.3)
- wait for `cluster-monitoring` to be active
- upgrade the app `monitoring-operator` up to date (v0.0.3)
- go to the project `p1`
- upgrade the app `project-monitoring` up to date (v0.0.3)

Result
- upgrading is successful and both cluster monitoring and project monitoring still work


Test 2:
- run Rancher: v2.2.3 container
- add an RKE EC2 cluster
- enable cluster monitoring 
- upgrade Rancher to v2.2.4-rc6
- go to project `system` - `Apps`
- upgrade the app `cluster-monitoring` up to date (v0.0.3)
- wait for `cluster-monitoring` to be active
- upgrade the app `monitoring-operator` up to date (v0.0.3)

Result:
- upgrading is successful and both cluster monitoring still work
